### DISCUSSION TITLE
request-reply subscribe to "range" or "set" instead of wildcard

### DISCUSSION THREAD

**Discussion by EgusCpp** [nats-io/nats-server#4911] [Q&A]
Greetings!

Let's say we have a command addressed by different clients. And for each set of clients we have preferrable server. So, by default, each server may subscribe to each subject separately per client ID, e.g. call one by one "cmd.001", "cmd.145", "cmd.222" and so on. Is it possible to subscribe it in one call, like first server subcribes to "cmd.001-200", another to "cmd.201-400" etc. Or, ideally, to submit a set of id numbers for routing. Is it ever possible to set up such kind of routing?  
Each server has preloaded clients data to reach better performance, i.e. it would be expensive to synchronize clients data in the case if every server may serve any client. At the same time, request-reply request is preferrable because it gives very quick response in the case when there are no subcsribers for requested subject.  
  
P.S. of course there is an option to make clients know which server they are addressing, while it doesn't look as good idea. It would be great to let clients to submit commands to NATS without knowing anything about command executors behind it.

**Comment by derekcollison** [comment]
The design of NATS is such that a client application can connect to any server within the system and function properly.

**Reply by EgusCpp** [reply]
In other words, there is no routing rules for request-reply processing inside NATS, so if we need to process part of requests from component A on server B, and another part on server C, depending on some request parameter (like client ID mentioned above), there is only way to make each of them subscribe all subjects one by one? I.e. server B is subscribing to set of 200 separate subjects (in the case if it serves 200 "clients").

**Reply by derekcollison** [reply]
NATS using same routing based in interest graph for request/reply. And the repliers can form a dynamic group where only one responder will receive any given request (see microservices framework in your NATS client).

**Reply by EgusCpp** [reply]
I have to apologize, I made unclear description, so looks like we're speaking about different things. Please let me show simple diagram:  
![image](https://github.com/nats-io/nats-server/assets/115817065/42ce6b71-a6e9-434b-ba9f-9d210e184f1d)

I am using NATS.c client library, so function names are from there. When server A needs to do something for client, it calls
**natsConnection_RequestString**  having subject with name made of "cmd" word and clientID, e.g. "cmd.1", "cmd.200" , etc.  

If we had only Server B for replying, we were able to subscribe using wildcard: "cmd.*" and serve all.  

While the issue is that we have more than one server, where each one could serve only its own set of clients.
E.g. if we have 10000 clients per each, then serverB and serverC needs to call **natsConnection_Subscribe** 10000 times. It sounds ridiculous. As alternative, we may make server A know about topology behind NATS, and keep some kind of addressing map internally, to call something like "cmd.serverB" and "cmd.serverC", which is also not good. That's why I was asking if there an alternative to simple wildcard, where instead of calling "cmd.*", we may define some range for single subscription call (like "cmd.1-200") or set (like "cmd.1,3,5,17,99").

**Reply by derekcollison** [reply]
In cases like this you can manually shard the clients with an additional token. So say Server A handle all even client IDs, and Server B handles All odd client IDs. The NATS system can also use automatic partitioning on the client's behalf.

https://docs.nats.io/nats-concepts/subject_mapping
https://docs.nats.io/nats-concepts/subject_mapping#deterministic-subject-token-partitioning

You can combine that technique with queue groups create compelling solutions.

**Reply by EgusCpp** [reply]
I see, thank you!

So, according to subject mapping, rule how to distribute messages is being set by NATS configuration, and repliers cannot dictate what subset of messages they would prefer to serve. And when Server B or Server C wants to know what data to preload, have to request a set of test messages with each possible ID, to see which of them will be returned via wilcard subscription...

**Reply by derekcollison** [reply]
If there are semantics that make sense around groupings you can put that as an extra token in a subject directly.
